asserts/ifacedecls.go

```diff
@@ -220,7 +222,263 @@ func compileAttributeConstraints(constraints interface{}) (*AttributeConstraints
+		panic("unknown PlugConnectionConstraints field " + field)
+	}
+}
+
+func (c *PlugConnectionConstraints) setIDConstraints(field string, cstrs []string) {
+	switch field {
+	case "slot-snap-type":
+		c.SlotSnapTypes = cstrs
+	case "slot-snap-id":
+		c.SlotSnapIDs = cstrs
+	case "slot-publisher-id":
+		c.SlotPublisherIDs = cstrs
+	default:
+		panic("unknown PlugConnectionConstraints field " + field)
+	}
+}
+
+var (
+	attributeConstraints = []string{"plug-attributes", "slot-attributes"}
+	plugIDConstraints    = []string{"slot-snap-type", "slot-publisher-id", "slot-snap-id"}
+
+	validPlugIDConstraints = map[string]*regexp.Regexp{
+		"slot-snap-type":    regexp.MustCompile("^(?:core|kernel|gadget|app)$"),
+		"slot-snap-id":      regexp.MustCompile("^[a-z0-9A-Z]{32}$"),                                             // snap-ids look like this
+		"slot-publisher-id": regexp.MustCompile("^(?:[a-z0-9A-Z]{32}|[-a-z0-9]{2,28}|\\$[a-z](?:-?[a-z0-9])*)$"), // account ids look like snap-ids or are nice identifiers, support our own special markers $MARKER
+	}
+)
+
+func compilePlugConnectionConstraints(interfaceName string, entry string, constraints interface{}) (*PlugConnectionConstraints, error) {
+	context := fmt.Sprintf("%s in plug rule for interface %q", entry, interfaceName)
+	cMap, shortcut, err := checkMapOrShortcut(context, constraints)
+	if err != nil {
+		return nil, err
+	}
+	if cMap == nil {
+		if shortcut == "true" {
+			return &PlugConnectionConstraints{PlugAttributes: AlwaysMatchAttributes, SlotAttributes: AlwaysMatchAttributes}, nil
+		}
+		// "false"
+		return &PlugConnectionConstraints{PlugAttributes: NeverMatchAttributes, SlotAttributes: NeverMatchAttributes}, nil
+	}
+	plugConnCstrs := &PlugConnectionConstraints{}
+	defaultUsed := 0
+	for _, field := range plugIDConstraints {
+		l, err := checkStringListInMap(cMap, field, fmt.Sprintf("%s in %s", field, context), validPlugIDConstraints[field])
+		if err != nil {
+			return nil, err
+		}
+		if l == nil {
+			defaultUsed++
+		}
+		plugConnCstrs.setIDConstraints(field, l)
+	}
+	for _, field := range attributeConstraints {
+		cstrs := AlwaysMatchAttributes
+		v := cMap[field]
+		if v != nil {
+			var err error
+			cstrs, err = compileAttributeConstraints(cMap[field])
+			if err != nil {
+				return nil, fmt.Errorf("cannot compile %s in %s: %v", field, context, err)
+			}
+		} else {
+			defaultUsed++
+		}
+		plugConnCstrs.setAttributeConstraints(field, cstrs)
+	}
+	if defaultUsed == len(attributeConstraints)+len(plugIDConstraints) {
+		return nil, fmt.Errorf("%s must specify at least one of %s, %s", context, strings.Join(attributeConstraints, ", "), strings.Join(plugIDConstraints, ", "))
+	}
+	return plugConnCstrs, nil
+}
+
+var (
+	defaultOutcome = map[string]interface{}{
+		"allow-installation":    "true",
+		"allow-connection":      "true",
+		"allow-auto-connection": "true",
+		"deny-installation":     "false",
+		"deny-connection":       "false",
+		"deny-auto-connection":  "false",
+	}
```

Comment:
I don't know myself either, anyway that's why I coded this way very declaritvely/explicitly, we can also have different approaches on the slots vs plugs side if needed

the other thing we were discussing with Gustavo, is how much we want to move to have entries for everything in the base declaration